core/vm/jit.go

```diff
@@ -0,0 +1,592 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"fmt"
+	"math/big"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/hashicorp/golang-lru"
+)
+
+type progStatus int32
+
+const (
+	progUnknown progStatus = iota
+	progCompile
+	progReady
+	progError
+)
+
+var programs *lru.Cache
+
+func init() {
+	programs, _ = lru.New(defaultJitMaxCache)
+}
+
+// SetJITCacheSize recreates the program cache with the max given size. Setting
+// a new cache is **not** thread safe. Use with caution.
+func SetJITCacheSize(size int) {
+	programs, _ = lru.New(size)
+}
+
+// GetProgram returns the program by id or nil when non-existent
+func GetProgram(id common.Hash) *Program {
+	if p, ok := programs.Get(id); ok {
+		return p.(*Program)
+	}
+
+	return nil
+}
+
+// GenProgramStatus returns the status of the given program id
+func GetProgramStatus(id common.Hash) progStatus {
+	program := GetProgram(id)
+	if program != nil {
+		return progStatus(atomic.LoadInt32(&program.status))
+	}
+
+	return progUnknown
+}
+
+// Program is a compiled program for the JIT VM and holds all required for
+// running a compiled JIT program.
+type Program struct {
+	Id     common.Hash // Id of the program
+	status int32       // status should be accessed atomically
+
+	context *Context
+
+	instructions []instruction       // instruction set
+	mapping      map[uint64]int      // real PC mapping to array indices
+	destinations map[uint64]struct{} // cached jump destinations
+
+	code []byte
+}
+
+func NewProgram(code []byte) *Program {
+	program := &Program{
+		Id:           crypto.Sha3Hash(code),
+		mapping:      make(map[uint64]int),
+		destinations: make(map[uint64]struct{}),
+		code:         code,
+	}
+
+	programs.Add(program.Id, program)
+	return program
+}
+
+func (p *Program) addInstr(op OpCode, pc uint64, fn instrFn, data *big.Int) {
+	// PUSH and DUP are a bit special. They all cost the same but we do want to have checking on stack push limit
+	// PUSH is also allowed to calculate the same price for all PUSHes
+	// DUP requirements are handled elsewhere (except for the stack limit check)
+	baseOp := op
+	if op >= PUSH1 && op <= PUSH32 {
+		baseOp = PUSH1
+	}
+	if op >= DUP1 && op <= DUP16 {
+		baseOp = DUP1
+	}
+	base := _baseCheck[baseOp]
+	instr := instruction{op, pc, fn, nil, data, base.gas, base.stackPop, base.stackPush}
+
+	p.instructions = append(p.instructions, instr)
```

Comment:
Right, though we could still pre-alloc the slice based on code length if it improves performance.